### PR TITLE
Tests for log messages

### LINKED ISSUE #26: Tests for logging
It'd be nice to have a way to test logged statements as they are shown to the user. Mixing in the Logging trait like we do now makes it hard to capture logs or mock the logger. From what I can tell, there are a couple of ways to capture logging:

- Inject the logger in the constructor and mock it away in tests. Maybe even with a default constructor like `class Foo(logger: Logger = LogFactory.getLogger())` so it doesn't require a lot of extra code in the application. Downside to this is it's not as clean as adding a trait, and (I think) can't be used on objects. Not the biggest problems.
- Capture Logback output using a custom Logback appender. Similar to [Log4s](https://github.com/Log4s/log4s#log4s-testing). I'm not very familiar with Logback, so this could be a bit of work. 
- Capturing stdout

Logged statements should be able to be filtered on the right class. And shouldn't get confused when tests are ran multithreaded (not sure how 2 & 3 would work with this).

### PR DESCRIPTION
**Pull request by legopiraat**
Fixes #26 

Had to make a custom Appender which is then configured in the logback-test.xml. This enabled us to capter the log events and write that to a list which sadly has to be mutable. 

After that, a custom scalatest matcher makes it easy for us to test if the log messages are logged on a specific log level.

### CODE REVIEW

File: util/src/test/scala/stryker4s/scalatest/LogMatchers.scala [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        s"Log message $expectedLogMessage was expected to be logged as $expectedLogLevel but wasn't.",
+        s"Log message $expectedLogMessage was logged as $expectedLogLevel."
+      )
+    }
+
+    private def validateLogLevel(actualLogLevel: Level): Boolean = {
+      expectedLogLevel.equals(actualLogLevel)
+    }
+
+    private def logMessageExists(expectedLogMessage: String): Boolean = {
```
[the remark this answers is not in the document]
Comment (legopiraat) [reply util/src/test/scala/stryker4s/scalatest/LogMatchers.scala]: Done and actually made the log level check work. Didn't even see i was comparing the same objects XD

File: util/src/test/scala/stryker4s/scalatest/LogMatchers.scala [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    }
+  }
+
+  class LogMatcherWithLevel(expectedLogLevel: Level) extends BeMatcher[String] {
+    def apply(expectedLogMessage: String): MatchResult = {
+      val result = logMessageExists(expectedLogMessage) && validateLogLevel(expectedLogLevel)
+
+      MatchResult(
+        result,
+        s"Log message $expectedLogMessage was expected to be logged as $expectedLogLevel but wasn't.",
```
[the remark this answers is not in the document]
Comment (legopiraat) [reply util/src/test/scala/stryker4s/scalatest/LogMatchers.scala]: Done

File: util/src/test/scala/stryker4s/TestAppender.scala [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+import ch.qos.logback.core.AppenderBase
+
+object TestAppender {
+  var events: List[ILoggingEvent] = List()
+}
+
+class TestAppender extends AppenderBase[ILoggingEvent] {
+
+  override def append(eventObject: ILoggingEvent): Unit =
+    TestAppender.events = TestAppender.events :+ eventObject
```
[the remark this answers is not in the document]
Comment (legopiraat) [reply util/src/test/scala/stryker4s/TestAppender.scala]: Ahh right dind't think of that

**Comment by legopiraat** [comment]
Added a reset function now so after each test the reset function can be called so it won't leak messages to other tests.

**Comment by hugo-vrijswijk** [comment]
I've run Stryker4s on itself, to see what the score would now be. A lot of undetected mutations come from logging statements, which we can now test!

https://gist.github.com/hugo-vrijswijk/4f93fdb12325b615329c00ccf333877c

**Pull request merged**